### DISCUSSION TITLE
Release 1.10.0 of the Amazon Kinesis Video WebRTC C SDK

### DISCUSSION THREAD

**Discussion by disa6302** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#1949] [Announcements]
What’s new:

* Improved time-to-first-frame latency with new CMAKE flag ENABLE_KVS_THREADPOOL. This flag is Enabled by default.  KVS_USE_SIGNALING_CHANNEL_THREADPOOL  has been deprecated now. (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1806)
* Added feature to send master time-to-first-frame for JS viewer timeline visualization; enable with enableSendingMetricsToViewerViaDc (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1846)
* Optimized time taken to gather TURN servers using GetIceServerConfig by supporting an async call pattern. In case of srflx based peer connection, this is a latency improvement since the SDK does not unnecessarily wait on getting TURN server URLs during signaling  setup. ENABLE_KVS_THREADPOOL required (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1854)
* Changed default build type from Debug to Release; use -CMAKE_BUILD_TYPE=Debug for debug mode. (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1897)

Sample bug fixes:

* Sending key frames before other frames in Samples. Frames are now sent only after SRTP is READY (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1833)
* Added checks for empty/null session token (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1874)

Bug fixes / Improvements:

* Fixed trickle ICE to be a session level and media level attribute instead of it being only a media level attribute (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1813)
* Optimized time taken for STUN DNS resolution to have STUN server address ready for srflx candidate gathering - ENABLE_KVS_THREADPOOL required (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1812)
* Improved DTLS handshake time by moving to a state transition based approach to be more responsive to the incoming DTLS handshake messages - ENABLE_KVS_THREADPOOL required (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1824)
* Optimized ICE and TURN state machine state transitions. They allow early state transitions if possible, instead of waiting for next timer iteration (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1825)
*  Fixed stringop-overflow build issue with MBedTLS on GCC11 (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1823)
* Increasing timeout values on TURN allocation, create permission, and bind channel steps. This change is expected to improve connection establishment  success rates in slower networks  (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1896)
* Fixed signaling cache file creation failure issue on Windows (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1935)
* Added filtering-out of non-operation network interfaces in Windows (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1937)

Version bumps:

* Updated LWS version from 4.2.2 to 4.3.3 (https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1820)
* Updated producer C version from 1.5.1 to 1.5.2 ([v1.5.2](https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/releases/tag/v1.5.2))

Known issues:

* MBedTLS build fails on Mac M1 with current version

<hr /><em>This discussion was created from the release <a href='https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/releases/tag/v1.10.0'>Release 1.10.0 of the Amazon Kinesis Video WebRTC C SDK</a>.</em>

**Comment by mehmetcy84** [comment]
Hi @disa6302 
Using latest master, v1.10.1, commitID: f1073439213db4aa4b4b7cc1423609d920dd5489
On Windows, I am trying to build using the standart procedure I used for 1.9.1, following the directives in https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/.github/workflows/ci.yml

The libwebsocket build step fails with this error:

-- Build files have been written to: E:/webrtc110/open-source/libwebsockets/build/src/project_libwebsockets-build
[ 66%] Performing build step for 'project_libwebsockets'

Microsoft (R) Program Maintenance Utility Version 14.16.27048.0
Copyright (C) Microsoft Corporation.  All rights reserved.

[  0%] Creating build include dir
[  0%] Built target GENHDR
[  1%] Building C object lib/CMakeFiles/websockets.dir/plat/windows/windows-fds.c.obj
windows-fds.c
[  1%] Building C object lib/CMakeFiles/websockets.dir/plat/windows/windows-file.c.obj
windows-file.c
[  2%] Building C object lib/CMakeFiles/websockets.dir/plat/windows/windows-init.c.obj
windows-init.c
[  3%] Building C object lib/CMakeFiles/websockets.dir/plat/windows/windows-misc.c.obj
windows-misc.c
[  3%] Building C object lib/CMakeFiles/websockets.dir/plat/windows/windows-pipe.c.obj
windows-pipe.c
[  4%] Building C object lib/CMakeFiles/websockets.dir/plat/windows/windows-plugins.c.obj
windows-plugins.c
[  5%] Building C object lib/CMakeFiles/websockets.dir/plat/windows/windows-service.c.obj
windows-service.c
[  5%] Building C object lib/CMakeFiles/websockets.dir/plat/windows/windows-sockets.c.obj
windows-sockets.c
[  6%] Building C object lib/CMakeFiles/websockets.dir/__/win32port/win32helpers/gettimeofday.c.obj
gettimeofday.c
[  7%] Building C object lib/CMakeFiles/websockets.dir/tls/tls.c.obj
tls.c
[  7%] Building C object lib/CMakeFiles/websockets.dir/tls/tls-network.c.obj
tls-network.c
[  8%] Building C object lib/CMakeFiles/websockets.dir/tls/tls-sessions.c.obj
tls-sessions.c
[  9%] Building C object lib/CMakeFiles/websockets.dir/tls/openssl/openssl-tls.c.obj
openssl-tls.c
[  9%] Building C object lib/CMakeFiles/websockets.dir/tls/openssl/openssl-x509.c.obj
openssl-x509.c
[ 10%] Building C object lib/CMakeFiles/websockets.dir/tls/openssl/openssl-ssl.c.obj
openssl-ssl.c
[ 11%] Building C object lib/CMakeFiles/websockets.dir/tls/openssl/openssl-session.c.obj
openssl-session.c
[ 11%] Building C object lib/CMakeFiles/websockets.dir/tls/tls-client.c.obj
tls-client.c
[ 12%] Building C object lib/CMakeFiles/websockets.dir/tls/openssl/openssl-client.c.obj
openssl-client.c
[ 12%] Building C object lib/CMakeFiles/websockets.dir/core/alloc.c.obj
alloc.c
[ 13%] Building C object lib/CMakeFiles/websockets.dir/core/buflist.c.obj
buflist.c
[ 14%] Building C object lib/CMakeFiles/websockets.dir/core/context.c.obj
context.c
[ 14%] Building C object lib/CMakeFiles/websockets.dir/core/lws_dll2.c.obj
lws_dll2.c
[ 15%] Building C object lib/CMakeFiles/websockets.dir/core/lws_map.c.obj
lws_map.c
[ 16%] Building C object lib/CMakeFiles/websockets.dir/core/libwebsockets.c.obj
libwebsockets.c
[ 16%] Building C object lib/CMakeFiles/websockets.dir/core/logs.c.obj
logs.c
[ 17%] Building C object lib/CMakeFiles/websockets.dir/core/vfs.c.obj
vfs.c
[ 18%] Building C object lib/CMakeFiles/websockets.dir/misc/base64-decode.c.obj
base64-decode.c
[ 18%] Building C object lib/CMakeFiles/websockets.dir/misc/prng.c.obj
prng.c
[ 19%] Building C object lib/CMakeFiles/websockets.dir/misc/lws-ring.c.obj
lws-ring.c
[ 20%] Building C object lib/CMakeFiles/websockets.dir/misc/cache-ttl/lws-cache-ttl.c.obj
lws-cache-ttl.c
[ 20%] Building C object lib/CMakeFiles/websockets.dir/misc/cache-ttl/heap.c.obj
heap.c
[ 21%] Building C object lib/CMakeFiles/websockets.dir/misc/cache-ttl/file.c.obj
file.c
[ 22%] Building C object lib/CMakeFiles/websockets.dir/misc/dir.c.obj
dir.c
[ 22%] Building C object lib/CMakeFiles/websockets.dir/misc/lwsac/lwsac.c.obj
lwsac.c
[ 23%] Building C object lib/CMakeFiles/websockets.dir/misc/lwsac/cached-file.c.obj
cached-file.c
[ 24%] Building C object lib/CMakeFiles/websockets.dir/misc/sha-1.c.obj
sha-1.c
[ 24%] Building C object lib/CMakeFiles/websockets.dir/misc/lejp.c.obj
lejp.c
[ 25%] Building C object lib/CMakeFiles/websockets.dir/system/system.c.obj
system.c
[ 25%] Building C object lib/CMakeFiles/websockets.dir/system/smd/smd.c.obj
smd.c
[ 26%] Building C object lib/CMakeFiles/websockets.dir/core-net/dummy-callback.c.obj
dummy-callback.c
[ 27%] Building C object lib/CMakeFiles/websockets.dir/core-net/output.c.obj
output.c
[ 27%] Building C object lib/CMakeFiles/websockets.dir/core-net/close.c.obj
close.c
[ 28%] Building C object lib/CMakeFiles/websockets.dir/core-net/network.c.obj
network.c
[ 29%] Building C object lib/CMakeFiles/websockets.dir/core-net/vhost.c.obj
vhost.c
[ 29%] Building C object lib/CMakeFiles/websockets.dir/core-net/pollfd.c.obj
pollfd.c
[ 30%] Building C object lib/CMakeFiles/websockets.dir/core-net/service.c.obj
service.c
[ 31%] Building C object lib/CMakeFiles/websockets.dir/core-net/sorted-usec-list.c.obj
sorted-usec-list.c
[ 31%] Building C object lib/CMakeFiles/websockets.dir/core-net/wsi.c.obj
wsi.c
[ 32%] Building C object lib/CMakeFiles/websockets.dir/core-net/wsi-timeout.c.obj
wsi-timeout.c
[ 33%] Building C object lib/CMakeFiles/websockets.dir/core-net/adopt.c.obj
adopt.c
[ 33%] Building C object lib/CMakeFiles/websockets.dir/roles/pipe/ops-pipe.c.obj
ops-pipe.c
[ 34%] Building C object lib/CMakeFiles/websockets.dir/core-net/state.c.obj
state.c
[ 35%] Building C object lib/CMakeFiles/websockets.dir/core-net/client/client.c.obj
client.c
[ 35%] Building C object lib/CMakeFiles/websockets.dir/core-net/client/connect.c.obj
connect.c
[ 36%] Building C object lib/CMakeFiles/websockets.dir/core-net/client/connect2.c.obj
connect2.c
[ 37%] Building C object lib/CMakeFiles/websockets.dir/core-net/client/connect3.c.obj
connect3.c
[ 37%] Building C object lib/CMakeFiles/websockets.dir/core-net/client/connect4.c.obj
connect4.c
[ 38%] Building C object lib/CMakeFiles/websockets.dir/core-net/client/sort-dns.c.obj
sort-dns.c
[ 38%] Building C object lib/CMakeFiles/websockets.dir/core-net/client/conmon.c.obj
conmon.c
[ 39%] Building C object lib/CMakeFiles/websockets.dir/roles/http/header.c.obj
header.c
[ 40%] Building C object lib/CMakeFiles/websockets.dir/roles/http/date.c.obj
date.c
[ 40%] Building C object lib/CMakeFiles/websockets.dir/roles/http/parsers.c.obj
parsers.c
[ 41%] Building C object lib/CMakeFiles/websockets.dir/roles/http/cookie.c.obj
cookie.c
[ 42%] Building C object lib/CMakeFiles/websockets.dir/roles/http/server/lejp-conf.c.obj
lejp-conf.c
[ 42%] Building C object lib/CMakeFiles/websockets.dir/roles/h1/ops-h1.c.obj
ops-h1.c
[ 43%] Building C object lib/CMakeFiles/websockets.dir/roles/h2/http2.c.obj
http2.c
[ 44%] Building C object lib/CMakeFiles/websockets.dir/roles/h2/hpack.c.obj
hpack.c
[ 44%] Building C object lib/CMakeFiles/websockets.dir/roles/h2/ops-h2.c.obj
ops-h2.c
[ 45%] Building C object lib/CMakeFiles/websockets.dir/roles/ws/ops-ws.c.obj
ops-ws.c
[ 46%] Building C object lib/CMakeFiles/websockets.dir/roles/ws/client-ws.c.obj
client-ws.c
[ 46%] Building C object lib/CMakeFiles/websockets.dir/roles/ws/client-parser-ws.c.obj
client-parser-ws.c
E:\webrtc110\open-source\libwebsockets\build\src\project_libwebsockets\lib\roles\ws\client-parser-ws.c(480): error C2220: warning treated as error - no 'object' file generated
E:\webrtc110\open-source\libwebsockets\build\src\project_libwebsockets\lib\roles\ws\client-parser-ws.c(480): warning C4003: not enough arguments for function-like macro invocation 'lwsl_hexdump_wsi'
E:\webrtc110\open-source\libwebsockets\build\src\project_libwebsockets\lib\roles\ws\client-parser-ws.c(480): error C2059: syntax error: ')'
E:\webrtc110\open-source\libwebsockets\build\src\project_libwebsockets\lib\roles\ws\client-parser-ws.c(598): warning C4003: not enough arguments for function-like macro invocation 'lwsl_hexdump_wsi'
E:\webrtc110\open-source\libwebsockets\build\src\project_libwebsockets\lib\roles\ws\client-parser-ws.c(598): error C2059: syntax error: ')'
NMAKE : fatal error U1077: '"C:\Program Files\CMake\bin\cmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
CMake Error at CMake/Utilities.cmake:75 (message):
  CMake step for libwebsockets failed: 2
Call Stack (most recent call first):
  CMakeLists.txt:193 (build_dependency)


-- Configuring incomplete, errors occurred!
See also "E:/webrtc110/build/CMakeFiles/CMakeOutput.log".

E:\webrtc110\build>nmake

Microsoft (R) Program Maintenance Utility Version 14.16.27048.0
Copyright (C) Microsoft Corporation.  All rights reserved.

NMAKE : fatal error U1064: MAKEFILE not found and no target specified
Stop.

I have tried setting -DLWS_WITH_MINIMAL_EXAMPLES=0 in 
E:\webrtc110\CMake\Dependencies\libwebsockets-CMakeLists.txt
but I get the same error.

**Comment by disa6302** [comment]
Can you confirm which visual studio version is being used? We updated the lws version to 4.3.3 in the recent version and our CI passes. Can you confirm the differences between our CI and your set up?

**Reply by mehmetcy84** [reply]
E:\webrtc110\build>cmd.exe /c cmake -G "NMake Makefiles" ..
-- Kinesis Video WebRTC Client path is E:/webrtc110
-- dependencies install path is E:/webrtc110/open-source
-- Begin building dependencies.
-- Copying patches for dependency openssl:
-- The C compiler identification is MSVC 19.16.27048.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- PREFIX is E:/webrtc110/open-source/libopenssl/build
-- CONFIGURE_COMMAND is perl;E:/webrtc110/open-source/libopenssl/build/src/project_libopenssl/Configure;VC-WIN64A;no-asm;--prefix=E:/webrtc110/open-source;--openssldir=E:/webrtc110/open-source
-- BUILD_COMMAND is C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx64/x64/nmake.exe
-- INSTALL_COMMAND is C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx64/x64/nmake.exe install_sw
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    BUILD_OPENSSL_PLATFORM
    BUILD_STATIC_LIBS
    OPENSSL_EXTRA

**Reply by disa6302** [reply]
Ok. I wonder if newer version does not work with version 2017. Based on https://github.com/warmcat/libwebsockets/issues/2635, the issue could be with older MSVC version. Could you try updating it to see if it helps?

**Reply by mehmetcy84** [reply]
Tested the 1.10.1 with VS2022 and build succeeded.
